### PR TITLE
Add Service Definitions Create and List APIs

### PR DESCRIPTION
**Pull request by himanshuvar**
**What this PR does / why we need it**:
This PR will do the following:

- Implement GET, LIST, POST for Service Definitions.
- Add Service Definition and Workflow Definitions in the database.                                                                                                                                
- Add Model Associations for service and workflow definitions.                                                                                                               
- Add Filters, Joins for the service and workflow definitions.

**APIs Output:**

**POST /v1beta/orchestration/services**
**Request Body:**
 ```
{
	"name":"volume povision",
	"description":"volume provisioning service",
	"tenant_id": "7fc5d8d4e24943a0967d9479b0c43cef",
        "user_id": "648e308959784c649ef0fffa7aa047a1",
        "input": "",
        "constraint": "",
        "group":"provisioning",
       "workflows":  [
                {
                    "definition_source": "opensds.provision-volume",
                    "wfe_type": "st2"
                }
    ]
}
```
**Response:**
```
{
    "constraint": "",
    "created_at": "Wed, 08 May 2019 21:41:24 GMT",
    "description": "volume provisioning service",
    "group": "provisioning",
    "id": "c3705fdd-ca63-48c6-9006-2944f7f37de6",
    "input": "",
    "name": "volume povision",
    "tenant_id": "7fc5d8d4e24943a0967d9479b0c43cef",
    "updated_at": "Wed, 08 May 2019 21:41:24 GMT",
    "user_id": "648e308959784c649ef0fffa7aa047a1"
}
```

**GET /v1beta/orchestration/services**
**Response:**
```
{
    "services": [
        {
            "service": {
                "constraint": "",
                "created_at": "Wed, 08 May 2019 21:41:24 GMT",
                "description": "s3 bucket migration service",
                "group": "migration",
                "id": "4ddc5609-afe5-4e82-a614-8fb0765aaa49",
                "input": "",
                "name": "bucket migration",
                "tenant_id": "7fc5d8d4e24943a0967d9479b0c43cef",
                "updated_at": "Wed, 08 May 2019 21:41:24 GMT",
                "user_id": "648e308959784c649ef0fffa7aa047a1",
                "workflows": [
                    {
                        "definition": {
                            "auth": {
                                "description": "Authentication Token.",
                                "required": true,
                                "type": "string"
                            },
                            "description": {
                                "description": "Description about the Bucket Migration Dataflow.",
                                "required": true,
                                "type": "string"
                            },
                            "destBackend": {
                                "description": "Destination Backend Storage.",
                                "required": true,
                                "type": "string"
                            },
                            "destBucketName": {
                                "description": "Destination Bucket Name.",
                                "required": true,
                                "type": "string"
                            },
                            "hostIP": {
                                "description": "Host IP for the OpenSDS.",
                                "required": true,
                                "type": "string"
                            },
                            "name": {
                                "description": "Name for the Bucket Migration Dataflow.",
                                "required": true,
                                "type": "string"
                            },
                            "port": {
                                "description": "Port for the service.",
                                "required": true,
                                "type": "string"
                            },
                            "remainSource": {
                                "description": "Value for keeping the source objects.",
                                "required": true,
                                "type": "boolean"
                            },
                            "srcBucketName": {
                                "description": "Source Bucket Name.",
                                "required": true,
                                "type": "string"
                            },
                            "tenantId": {
                                "description": "Tenant ID.",
                                "required": true,
                                "type": "string"
                            },
                            "userId": {
                                "description": "User ID.",
                                "required": true,
                                "type": "string"
                            }
                        },
                        "description": "Bucket Migration Multi-Cloud Service",
                        "id": "85b13a76-985f-4414-abf1-d04466ac4209",
                        "name": "migration-bucket"
                    }
                ]
            }
        },
        {
            "service": {
                "constraint": "",
                "created_at": "Wed, 08 May 2019 21:41:24 GMT",
                "description": "volume provisioning service",
                "group": "provisioning",
                "id": "c3705fdd-ca63-48c6-9006-2944f7f37de6",
                "input": "",
                "name": "volume povision",
                "tenant_id": "7fc5d8d4e24943a0967d9479b0c43cef",
                "updated_at": "Wed, 08 May 2019 21:41:24 GMT",
                "user_id": "648e308959784c649ef0fffa7aa047a1",
                "workflows": [
                    {
                        "definition": {
                            "accessprotocol": {
                                "required": false,
                                "type": "string"
                            },
                            "availabilityzone": {
                                "required": false,
                                "type": "string"
                            },
                            "connectioninfo": {
                                "required": false,
                                "type": "object"
                            },
                            "description": {
                                "required": false,
                                "type": "string"
                            },
                            "hostinfo": {
                                "required": false,
                                "type": "object"
                            },
                            "ipaddr": {
                                "required": true,
                                "type": "string"
                            },
                            "mountpoint": {
                                "required": false,
                                "type": "string"
                            },
                            "name": {
                                "required": true,
                                "type": "string"
                            },
                            "port": {
                                "required": true,
                                "type": "string"
                            },
                            "profileid": {
                                "required": false,
                                "type": "string"
                            },
                            "projectid": {
                                "required": true,
                                "type": "string"
                            },
                            "size": {
                                "required": true,
                                "type": "integer"
                            },
                            "snapshotfromcloud": {
                                "required": false,
                                "type": "string"
                            },
                            "snapshotid": {
                                "required": false,
                                "type": "string"
                            },
                            "tenantid": {
                                "required": false,
                                "type": "string"
                            },
                            "timeout": {
                                "default": 60,
                                "type": "integer"
                            }
                        },
                        "description": "Provision an OpenSDS Volume",
                        "id": "b146f45d-fd7f-4e7f-b0de-ed34fccdcfa9",
                        "name": "provision-volume"
                    }
                ]
            }
        }
    ]
}
```

### CODE REVIEW

File: orchestration/api/services.py [hunk 1]
```diff
@@ -14,15 +14,114 @@
...
+        service_data["workflow_definitions"] = [wf_ob]
+        create_service_definition(None, service_data)
+
+    service_def_hash = get_service_def(service_data['id'])
+
+    return jsonify(response=service_def_hash), 200
+
+
+def get_service_def(service_id):
+    service_def_list = get_service_definition(service_id, None)
```
Comment (kumarashit) [inline orchestration/api/services.py]: Just check if return is None or []

File: orchestration/db/api.py [hunk 2]
```diff
@@ -58,7 +95,10 @@ def get_service_definition(id='', context=None):
         else:
             query = session.query(models.ServiceDefinition).filter(
                 models.ServiceDefinition.id == id)
-    return None if not query else query.first()
+    if not query:
+        return None
+    else:
+        return get_query_res(query.all(), models.ServiceDefinition)
```
Comment (kumarashit) [inline orchestration/db/api.py]: Why don't we use query.first()? Else, in the caller you will have to do res[0]

File: orchestration/api/services.py [hunk 3]
```diff
@@ -14,15 +14,114 @@
...
+    wfs = []
+    for sd, wd in service_workflow_list:
+        wfd_hash = {'id': wd.id,
+                    'name': wd.name,
+                    'description': wd.description,
+                    'definition': json.loads(wd.definition)
+                    }
+        wfs.append(wfd_hash)
+
+    sd_hash = service_def_list[0]
```
Comment (kumarashit) [inline orchestration/api/services.py]: if query returns query.first() then we may not need this i.e. getting first elem, right?

File: orchestration/api/services.py [hunk 4]
```diff
@@ -14,15 +14,114 @@
...
+                        }
+            # check if the entries are not present in DB then only
+            # enter in DB
+            if get_workflow_definition(None, elem['ref']) is None and \
+                    re.search(service_data["name"],
+                              elem['ref'], re.IGNORECASE):
+                wf_obj = create_workflow_definition(None, wfd_hash)
+                wfs.append(wfd_hash)
+                wf_objs.append(wf_obj)
+            wf_hash['workflows'] = wfs
```
Comment (kumarashit) [inline orchestration/api/services.py]: Is this used? wf_hash

File: orchestration/api/services.py [hunk 5]
```diff
@@ -14,15 +14,114 @@
 
 from flask import jsonify
```
Comment (kumarashit) [inline orchestration/api/services.py]: can we merge these three lines

File: orchestration/db/api.py (hunk 2 above)
Comment (himanshuvar) [reply orchestration/db/api.py]: In the caller, result is the list of the dict for the db objects.

File: orchestration/api/services.py (hunk 3 above)
Comment (himanshuvar) [reply orchestration/api/services.py]: We need to get the single(first) elem from the list.

File: orchestration/api/services.py (hunk 4 above)
Comment (himanshuvar) [reply orchestration/api/services.py]: Removed the unused code.

File: orchestration/api/services.py (hunk 5 above)
Comment (himanshuvar) [reply orchestration/api/services.py]: Yes, Merged

File: orchestration/db/api.py [hunk 6]
```diff
@@ -36,6 +37,42 @@ def session_scope():
         session.close()
 
 
+# -------------------------data access for service definition and
+#                           workflow definition associations------------------
```
Comment (xing-yang) [inline orchestration/db/api.py]: nit: this commenting style looks weird.  Usually there is "#----" in its own line in the beginning and "#----" in its own line at the end, but not mixed like this.  So it is easier to just remove ----

File: tests/unit/db/test_api.py [hunk 7]
```diff
@@ -39,6 +39,8 @@ def test_create_service_definition(mock_session, mock_uuid):
     mock_session.return_value.__enter__.return_value.add.assert_called_once
 
 
+"""
+TODO
```
Comment (xing-yang) [inline tests/unit/db/test_api.py]: This section is commented out.  Please remove it now and add back later when it is working.

File: orchestration/api/services.py (hunk 1 above)
Comment (wisererik) [reply orchestration/api/services.py]: get_service_definition only return an object or None, not the array

File: requirements.txt [hunk 8]
```diff
@@ -4,6 +4,6 @@ sqlalchemy
 st2client
 uuid
 cryptography
-requests
+requests<=2.15
 ConfigParser
 mock
```
Comment (wisererik) [inline requirements.txt]: remove mock in requirement.txt

Comment (himanshuvar) [reply requirements.txt]: Done

File: tests/unit/db/test_api.py (hunk 7 above)
Comment (himanshuvar) [reply tests/unit/db/test_api.py]: Done

File: orchestration/db/api.py (hunk 6 above)
Comment (himanshuvar) [reply orchestration/db/api.py]: Removed the lines.

File: orchestration/api/services.py (hunk 1 above)
Comment (himanshuvar) [reply orchestration/api/services.py]: Done. Added the error conditions

File: orchestration/db/models.py [hunk 9]
```diff
@@ -27,16 +28,21 @@ class ModelBase(Base):
     __abstract__ = True
     id = Column(String(36),  default=lambda: str(
         uuid.uuid4()), primary_key=True)
-    created_at = Column(DateTime)
-    updated_at = Column(DateTime)
+    created_at = Column(DateTime, default=datetime.datetime.utcnow())
```
Comment (wisererik) [inline orchestration/db/models.py]: add lambda expression here

Comment (himanshuvar) [reply orchestration/db/models.py]: Updated the PR with new changes

File: orchestration/connectionmanager/st2.py [hunk 10]
```diff
@@ -61,6 +61,17 @@ def list_actions(self, pack_name=''):
...
+    # Function to get the action
+    # Caller can store the result in DB
+    # @Input: PackName, ActionId
+    # @output: the ActionDetails
+    def get_action(self, id, pack_name=''):
+        auth_token = self.authenticate()
+        headers = {'X-Auth-Token': auth_token}
+        url = Orchconstants().get_st2_action_list_url(self.server) + '/' + id
+        response = requests.get(url, headers=headers, verify=False)
+        return response.json()
```
Comment (kumarashit) [inline orchestration/connectionmanager/st2.py]: Please return response.status_code, response.json()
In the caller just check if return code is != Apiconstants.HTTP_OK then return the error in response with the proper HTTP error code we got

Comment (himanshuvar) [reply orchestration/connectionmanager/st2.py]: Done

File: orchestration/db/api.py [hunk 11]
```diff
@@ -36,29 +37,70 @@ def session_scope():
...
+            values['id'] = str(uuid.uuid4())
+
+        service_definition = models.ServiceDefinition()
+        service_definition.update(values)
+        for workflow_definition in workflow_definitions:
+            service_definition.workflow_definitions.append(workflow_definition)
+
+        with session_scope() as session:
+            session.add(service_definition)
+    except RuntimeError:
```
Comment (kumarashit) [inline orchestration/db/api.py]: Ideally we should also handle SQLAlchemyError too

Comment (kumarashit) [reply orchestration/db/api.py]: Ok

**Pull request merged**